### ISSUE TITLE
false positives due to utf-8/german umlauts

### ISSUE THREAD

**Issue by thorstenhirsch** [AtomLinter/linter-ruby#69]
The linter throws "unexpected end-of-input, expecting keyword_end" errors for both 'ö' characters in this simple ruby class:

```ruby
class Foo
  flash = Hash.new
  flash[:alert] = 'Smörebröd'
end
```

Atom says that the encoding is UTF-8, the locales on my OS X system are also UTF-8.

**Comment by Arcanemagus** [comment]
Does `ruby -wc foo.rb` show the same errors? If so the problem lies within `ruby` and not this package.

**Comment by thorstenhirsch** [comment]
Hi Landon,

no, that command doesn't throw any errors:

$ ruby -wc foo.rb

Syntax OK

Regards,
Thorsten


2016-02-16 23:47 GMT+01:00 Landon Abney <notifications@github.com>:

**Comment by Arcanemagus** [comment]
Hmmm, I get the same results in the console:
![image](https://cloud.githubusercontent.com/assets/427137/13099848/977f4268-d4eb-11e5-8b84-a91518f385a0.png)

However, within Atom I get the correct results:
![image](https://cloud.githubusercontent.com/assets/427137/13099871/d08f4008-d4eb-11e5-9a1c-3f83f7e36afd.png)


You do say you are on OS X though, so what I'm guessing is going on is that you are getting the wrong version of `ruby` when it tries to execute it. If you place a breakpoint [here](https://github.com/AtomLinter/linter-ruby/blob/v1.2.1/lib/main.js#L44) do you see the incorrect result?

**Comment by thorstenhirsch** [comment]
Hi Landon,

I don't know how to insert a breakpoint, so I just added some console.log
commands and here's what I can see in the developer console:

-:114: invalid multibyte char (US-ASCII)
-:114: invalid multibyte char (US-ASCII)
-:114: syntax error, unexpected end-of-input, expecting keyword_end
          flash[:alert] = 'Smörebröd'

So it's definitely not trying to interpret the text in utf-8. I don't know
how or why, but the locales in Atom are not utf-8:

LANG=
LC_COLLATE="C"
LC_CTYPE="C"
LC_MESSAGES="C"
LC_MONETARY="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_ALL=

I temporarily set command='locales' in your linter's main.js to produce
this output. The locales in my user's shell are:

$ locale

LANG="de_DE.UTF-8"

LC_COLLATE="de_DE.UTF-8"

LC_CTYPE="de_DE.UTF-8"

LC_MESSAGES="de_DE.UTF-8"

LC_MONETARY="de_DE.UTF-8"

LC_NUMERIC="de_DE.UTF-8"

LC_TIME="de_DE.UTF-8"

LC_ALL=


I'm afraid the problem might be specific to OS X, because user environment
variables (including locales) are not available in the applications by
default. Well, at least it depends on how you start your applications. If
you start Atom in a shell, the forked process will of course inherit its
environment from the shell. But a common way is to use spotlight (that's
how I start Atom, too) and spotlight is started by launchctl (system), thus
there's no user environment available. It's really weird for users with
Windows or Linux background. Actually I wonder why no-one else had this
problem. There are quite a few developers on OS X out there.

Setting locales to de_DE.UTF-8 on system level fixes the issue.


But wouldn't that mean, that we always need to have locales that match the
file contents? Maybe you can reproduce the problem on Windows when using a
cp1252 environment with our "Smörebröd" utf-8 file...? Anyway, since the
linter depends on calling ruby, I see no way to get rid of relying on
locales that match the file contents ...or do you?


Regards

Thorsten



2016-02-17 5:29 GMT+01:00 Landon Abney <notifications@github.com>:

**Comment by Arcanemagus** [comment]
@thorstenhirsch Can you try version 1.2.2 and see if this is fixed?

**Comment by thorstenhirsch** [comment]
Yes, it is! Thank you!

**Comment by Arcanemagus** [comment]
Thank @ananevam for putting up the fix :wink:.
